ArduSub/mode_poshold.cpp

```diff
@@ -103,12 +83,42 @@ void ModePoshold::run()
             attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(target_roll, target_pitch, target_yaw_rate);
             sub.last_pilot_heading = ahrs.yaw_sensor; // update heading to hold
 
-        } else { // call attitude controller holding absolute absolute bearing
+        } else { // call attitude controller holding absolute bearing
             attitude_control->input_euler_angle_roll_pitch_yaw(target_roll, target_pitch, sub.last_pilot_heading, true);
         }
     }
 
-    // Update z axis //
+    // update z axis
     control_depth();
+
+    // update xy axis
+    // call this after Sub::get_pilot_desired_climb_rate is called so that THR_DZ is reasonable
+    control_horizontal();
+}
+
+void ModePoshold::control_horizontal() {
+    float lateral_out = 0;
+    float forward_out = 0;
+
+    if (sub.position_ok()) {
+        // get desired rates in the body frame
+        Vector2f body_rates_cm_s = {
+                sub.get_pilot_desired_horizontal_rate(channel_forward),
+                sub.get_pilot_desired_horizontal_rate(channel_lateral)
+        };
+
+        // convert to the earth frame and set target rates
+        auto earth_rates_cm_s = ahrs.body_to_earth2D(body_rates_cm_s);
+        position_control->input_vel_accel_xy(earth_rates_cm_s, {0, 0});
+
+        // convert pos control roll and pitch angles back to lateral and forward efforts
+        sub.translate_pos_control_rp(lateral_out, forward_out);
+
+        // udpate the xy controller
+        position_control->update_xy_controller();
+    }
+
+    motors.set_forward(forward_out);
```

Comment:
This looks like a good change to me because previously the pilot input was essentially added on top of the position controller output.  That would lead to the position controller fighting the user which is terrible and a pattern we try to avoid.

The change in behaviour for users is that they won't be able to control the vehicle laterally in poshold mode anymore unless the EKF has a good position estimate.

Still, looks like an improvement!

Reply:
> The change in behaviour for users is that they won't be able to control the vehicle laterally in poshold mode anymore unless the EKF has a good position estimate.

That's an interesting point. what if we move `body_rates_cm_s` up one level and apply that to thrusters if `posititon_ok()` is false? 

Sensors are not perfect and I think users should keep control when they fail.

Reply:
> That's an interesting point. what if we move `body_rates_cm_s` up one level and apply that to thrusters if `posititon_ok()` is false?

Ah, good catch, and good idea. Let me play around with this a bit.